packages/vite/src/node/preview.ts

```diff
@@ -129,16 +132,59 @@ export async function preview(
       }
     },
   })
-  app.use(previewBase, async (req, res, next) => {
-    if (shouldServe(req.url!, distDir)) {
-      return assetServer(req, res, next)
+  app.use(
+    previewBase,
+    // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+    async function vitePreviewStaticMiddleware(req, res, next) {
+      if (shouldServe(req.url!, distDir)) {
+        return assetServer(req, res, next)
+      }
+      next()
+    },
+  )
+
+  // html fallback
+  if (config.appType === 'spa' || config.appType === 'mpa') {
+    // append trailing slash when base didn't have it
+    if (config.rawBase !== config.base) {
+      // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+      app.use(function viteRewriteBaseAccessWithoutTrailingSlashMiddleware(
+        req,
+        res,
+        next,
+      ) {
+        try {
+          const pathname = decodeURIComponent(
+            new URL(req.url!, 'http://example.com').pathname,
+          )
+          if (pathname === config.rawBase) {
+            req.url = config.base + req.url!.slice(config.rawBase.length)
+          }
+        } catch {}
+        next()
+      })
     }
-    next()
-  })
+
+    app.use(
+      previewBase,
+      htmlFallbackMiddleware(distDir, config.appType === 'spa'),
+    )
```

Comment:
We could add a middleware here to support rewriting `/foo` to `/foo.html`. I didn't implement it because it wasn't supported in dev.

Reply:
Would it be possible to add an option in `preview` to [skip the `shouldServe` and let `sirv` handle it?](https://github.com/vitejs/vite/issues/10925#issuecomment-1344427220).

That would allow frameworks like îles to configure the `preview` to work in a way that is consistent with the deployment target (which might resolve `/foo` to `/foo.html`).

Reply:
I think that is possible by injecting a middleware by the [post `configurePreviewServer`](https://vitejs.dev/guide/api-plugin.html#configurepreviewserver) hook, too.
In that way, I think we can have the same behavior in dev by injecting that middleware by `configureServer` hook.